### PR TITLE
[libc++] <experimental/simd> Add assignment operator of simd reference

### PR DESCRIPTION
**Pull request by joy2myself**

### CODE REVIEW

File: libcxx/include/experimental/__simd/reference.h [hunk 1]
```diff
@@ -44,6 +44,13 @@ class __simd_reference {
   __simd_reference(const __simd_reference&) = delete;
 
   _LIBCPP_HIDE_FROM_ABI operator value_type() const noexcept { return __get(); }
+
+  template <class _Up>
+  _LIBCPP_HIDE_FROM_ABI std::enable_if_t<__reference_well_formed<value_type, _Up>, __simd_reference>
+  operator=(_Up&& __v) && noexcept {
```
Comment (philnik777) [inline libcxx/include/experimental/__simd/reference.h:50]:
```suggestion
  template <class _Up, enable_if_t<__reference_well_formed<value_type, _Up>, int> = 0>
  _LIBCPP_HIDE_FROM_ABI __simd_reference
  operator=(_Up&& __v) && noexcept {
```

File: libcxx/include/experimental/__simd/utility.h [hunk 2]
```diff
@@ -91,6 +91,14 @@ _LIBCPP_HIDE_FROM_ABI constexpr bool __can_generate(index_sequence<_Idxes...>) {
 template <class _Tp, class _Generator, std::size_t _Size>
 inline constexpr bool __can_generate_v = experimental::__can_generate<_Tp, _Generator>(make_index_sequence<_Size>());
 
+template <class _Tp, class _Up, class = void>
+inline constexpr bool __reference_well_formed = false;
+
+template <class _Tp, class _Up>
+inline constexpr bool
+    __reference_well_formed<_Tp, _Up, std::void_t<decltype(declval<_Tp&>() = std::forward<_Up>(declval<_Up&&>()))>> =
```
Comment (philnik777) [inline libcxx/include/experimental/__simd/utility.h:99]: This looks to me a lot like `is_assignable_v<_Tp&, _Up&&>`.

File: libcxx/include/experimental/__simd/reference.h [hunk 3]
```diff
@@ -44,6 +45,12 @@ class __simd_reference {
   __simd_reference(const __simd_reference&) = delete;
 
   _LIBCPP_HIDE_FROM_ABI operator value_type() const noexcept { return __get(); }
+
+  template <class _Up, std::enable_if_t<is_assignable_v<value_type&, _Up&&>, int> = 0>
```
Comment (philnik777) [inline libcxx/include/experimental/__simd/reference.h:49]:
```suggestion
  template <class _Up, enable_if_t<is_assignable_v<value_type&, _Up&&>, int> = 0>
```

File: libcxx/test/std/experimental/simd/simd.reference/reference_assignment.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+template <class T, class SimdAbi>
+struct CheckSimdReferenceAssignmentHelper {
+  template <class U>
+  void operator()() const {
+    if constexpr (std::is_assignable_v<T&, U&&>) {
+      ex::simd<T, SimdAbi> origin_simd([](T i) { return i; });
+      for (size_t i = 0; i < origin_simd.size(); ++i) {
+        origin_simd[i] = static_cast<U>(i + 1);
+        assert(origin_simd[i] == static_cast<T>(std::forward<U>(i + 1)));
```
Comment (philnik777) [inline libcxx/test/std/experimental/simd/simd.reference/reference_assignment.pass.cpp:29]: Please add a check to make sure `operator=` is `noexpect`.

**Review by philnik777 — APPROVED** [review]
LGTM with comments addressed.

**Comment by joy2myself** [comment]
Are these CI errors related? I didn't quite understand these error messages.

**Comment by philnik777** [comment]
The FreeBSD failure looks like a fluke. I have no idea what AIX is complaining about, but that one looks like it's related. I'd just add an `// XFAIL: LIBCXX-AIX-FIXME`. CC @DanielMcIntosh @perry-ca

**Comment by vitalybuka** [comment]
I've disabled the test with sanitizers, it timeouts on ARM

Also I just checked x86_64, it also slowest test:
```
533.59s: llvm-libc++-shared.cfg.in :: std/experimental/simd/simd.reference/reference_assignment.pass.cpp
150.11s: llvm-libc++-shared.cfg.in :: std/containers/sequences/deque/deque.modifiers/insert_size_value.pass.cpp
```
@philnik777 Maybe you want to tweak the test to make it faster (< 50s ? ) and reenable sanitizers?

**Comment by philnik777** [comment]
@vitalybuka What do you mean with "it timeouts on ARM"? Our CI ran just fine.

**Comment by vitalybuka** [comment]
> @vitalybuka What do you mean with "it timeouts on ARM"? Our CI ran just fine.

https://lab.llvm.org/buildbot/#/builders/239/builds/4263/steps/11/logs/stdio

How long it takes to run that test?
Can you run them with LIT_OPTS="--time-tests -v" ?

**Pull request merged**